### DISCUSSION TITLE
Mirth Filter for HL7 Messages

### DISCUSSION THREAD

**Discussion by jgh1189** [nextgenhealthcare/connect#5302] [Q&A]
I am creating a filter on the source side to remove any messages that have a PID-3.1 value that equals any ID in list of about 80k ID's that are in CSV file. I think my issue is getting Mirth to actually read the file, store the values or reference the values. 

I tried this from the other Mirth forum in post 4. https://forums.mirthproject.io/forum/mirth-connect/support/12427-filter-with-multiple-values  which is spot on to what I am trying to accomplish. 

I am not able to see the list was created and not able to see the global channel where the list should reside. When I run my test messages in with some that are in the exclude file and some that are not, the dashboard shows that the entire test file was filtered. The goal is remove the messages with the ID's in the CSV. 

Any help is greatly appreciated

**Comment by pacmano1** [comment]
Post your code. 

Also for that number of records I think most folks would be maintaining the list in a database and calling the database to check for a conflict or using Guava to build a cache.  ie. not using an internal memory map.  What volume of messages per hour are you processing?

**Reply by jgh1189** [reply]
I will look into the database option for the list of ID's that I need to exclude. For now I am testing with only about 10 values to see if I can get it to work. 

This is an already generated HL7 file that contains around 300k messages. These are ORU messages with multiple segments so the file itself is around 2.5 million rows. 

Under scripts:

var list = new java.util.ArrayList();
for each (id in FileUtil.read('/Mirth_Input/FileName.csv').split('\r\n|\r|\n')) {
	list.add(id.trim());
}
$gc('patientIds', list);

Under source>Edit Filter>Type is Javascript

return $gc('patientIds').contains(msg['PID']['PID.3']['PID.3.1'].toString()); //I understand this is contains but wanted to see if I could at least reference the list. 
//return $gc('patientIds')!=(msg['PID']['PID.3']['PID.3.1'].toString()); //Tried this as well for does not contain

**Comment by mjtrac** [comment]
I have essentially the same code as you, and it works fine.  If you want to see the list, you can just add a logger statement temporarily to log out the list after you retrieve it from the globalChannelMap.  Names.csv has three lines and no commas.
My code, deployment script:

`list = new java.util.ArrayList();`
`names = FileUtil.read("C:/Users/.../names.csv")`


` for each (x in names.split('\n')){` 
` 	list.add(x.trim());` 
` }` 
` globalChannelMap.put('names',list);` 
` return;`

And source filter:
`// require PID.5.1 to contain entry`
`logger.info(globalChannelMap.get('names'));`
`return (globalChannelMap.get('names').contains(msg['PID']['PID.5']['PID.5.1'].toString()));`

When a message is sent, it is filtered appropriately, and the following is printed to the dashboard and log:
 INFO  (filter:?): [one, two, three]

You might need to make sure your logging levels are set to display info messages, or use logger.error instead of logger info.

I have no idea whether this approach is wildly inefficient, but the code works and you see the ArrayList.  If you use the debugger introduced in 4.0, you'll be able to see the list via the debugger as well.

**Reply by tonygermano** [reply]
> I have no idea whether this approach is wildly inefficient

I'd definitely use a HashSet instead of an ArrayList. The ArrayList will need to iterate the entire list until it finds it or doesn't. The HashSet computes in constant time whether you have 10 records or 80k.

Should be able to do it like the following. You could initialize an ArrayList in exactly the same way by passing a js array (which is a `java.util.List`) to the constructor.
```js
var names = FileUtil.read("C:/Users/.../names.csv")
var list = new java.util.HashSet(names.split('\n').map(name => name.trim()))
globalChannelMap.put('names', list)
```
